modules/dmrpp_module/DmrppArray.cc

```diff
@@ -1373,25 +1374,74 @@ void DmrppArray::read_linked_blocks(){
-        
-        this->clear_local_data();
-        reserve_value_capacity_ll_byte(this->width_ll());
-        char *out_buf = get_buf();
-        memcpy(out_buf,dest_deflate,this->width_ll());
-        delete []dest_deflate; 
+        string filters_str = this->get_filters();
+        if (filters_str.find("deflate")!=string::npos) 
+            throw InternalErr(__FILE__, __LINE__, "We don't handle compressed array of structure now.");
+
+        // Check if we can handle this case. 
+        // Currently we only handle one-layer simple int/float types, and the data is not compressed. 
+        bool can_handle_struct = check_struct_handling();
+        if (!can_handle_struct) 
+            throw InternalErr(__FILE__, __LINE__, "Only handle integer and float base types. Cannot handle the array of complex structure yet."); 
+
+        vector<char> values;
+        values.resize(get_var_chunks_storage_size());
+        char *target_buffer = values.data();
+    
+        for(const auto& chunk: get_immutable_chunks()) {
+            chunk->read_chunk();
+            BESDEBUG(dmrpp_3, prolog << "linked_block_index: " << chunk->get_linked_block_index() << endl);
+            BESDEBUG(dmrpp_3, prolog << "accumlated_length: " << accumulated_lengths[chunk->get_linked_block_index()]  << endl);
+            const char *source_buffer = chunk->get_rbuf();
+            memcpy(target_buffer + accumulated_lengths[chunk->get_linked_block_index()], source_buffer,chunk->get_size());
+        }
+
+#if 0
+        char *buf_value = the_one_chunk->get_rbuf();
+        unsigned long long value_size = the_one_chunk->get_size();
+        vector<char> values(buf_value,buf_value+value_size);
+#endif
+        read_array_of_structure(values);
+    }
+    else {
+
+        //Change to the total storage buffer size to just the compressed buffer size.
+        reserve_value_capacity_ll_byte(get_var_chunks_storage_size());
+
+
+        char *target_buffer = get_buf();
+    
+        for(const auto& chunk: get_immutable_chunks()) {
+            chunk->read_chunk();
+            BESDEBUG(dmrpp_3, prolog << "linked_block_index: " << chunk->get_linked_block_index() << endl);
+            BESDEBUG(dmrpp_3, prolog << "accumlated_length: " << accumulated_lengths[chunk->get_linked_block_index()]  << endl);
+            const char *source_buffer = chunk->get_rbuf();
+            memcpy(target_buffer + accumulated_lengths[chunk->get_linked_block_index()], source_buffer,chunk->get_size());
+        }
+        string filters_string = this->get_filters();
+        if (filters_string.find("deflate")!=string::npos) {
+    
+            char *in_buf = get_buf();
+    
+            char **destp = nullptr;
+            char *dest_deflate = nullptr;
+            unsigned long long out_buf_size = 0;
+            unsigned long long dest_len = get_var_chunks_storage_size();
+            unsigned long long src_len = get_var_chunks_storage_size();
+            dest_deflate = new char[dest_len];
```

Comment:
For situations like this, you could use a unique_ptr<char[]> or a vector<char>. The benefit is that you do not have to delete (or delete[]) the memory yourself and thus have no leak if other code (e.g., reserve_value_capacity_ll_byte()) throws an exception.

For a vector '.data()' returns a pointer to the block of memory. Use .get() with a unique_ptr.

Reply:
This is the from the existing way and I recall vector was not working. I will leave it now and will check this in the future.